### PR TITLE
Song Creation form is not sufficiently user friendly #305

### LINKED ISSUE #14: Allow entering two or more "original" PVs on song submission page
Currently there's fields for one original PV and one reprint. Should add dropdowns for selecting type for each PV (by default the first can be original and second a reprint). Possibly a button for adding more PVs as well.

If more than one original PV is entered, how to load metadata (artist, song title)? First original Nico PV? What if the PV type is changed? Maybe there should always be one (fixed) field for original PV?

### LINKED ISSUE #305: Song Creation form is not sufficiently user friendly
Some frequent problems and misunderstandings happen due to mistakes in song creation (verbatim from @riipah):

> * Songs missing producers or Vocaloids (there is a warning for this)
> * Wrong Romanization (usually Romanizing loanwords - this should be more clearly on the edit page)
> * Cover/remix is missing mention of original (there is warning for this)
> * Unnecessary text in the title (usually this happens if the Nico title doesn't follow the common format and can't be automatically parsed, some people include "cover" or artists)
> * Covers that are added as originals (song type should be cover)

We can help users avoid these issues by changing the form's functionality as described below. The idea is to keep the old form around for experienced users, but make this one the default.

```
1. A group is available while there is no failure preceding it.
2. Checks are client-side.
3. Greyed out elements show why they are greyed out (icon with hover text).
4. Warning text should look scary

{}      Group (Visually distinct)
[]      Text input
< ()>   Drop-down (default option first, other options follow)
▣ □     Checkbox
◉ ○     Radio
(X)     Group-local selection identifier

*       "On selection of..."
%       "On field focus..."
@       "On form submission..."

!       "Fail while..."
?       "If"
->      "Do"

MSG     Display message
DRF     Force-enable {Draft}
REP     Submit report
```
```
{URL}

    [original PV]
    [reprint of the PV]
```
```
{Intro}

    ○ "Song has vocaloids singing in it" (A)
    ○ "Song does not have vocaloids singing in it" (B)
        □ "Song is part of an album that contains vocaloid songs"
        □ "Song is the Original to many vocaloid covers"

    /------------
    | * ? (B) -> MSG http://wiki.vocadb.net/wiki/39/vocadb-editing-faq
    | @ ? (B) -> DRF
    | ! (B) is the _only_ selection.
    \------------
```
```
{Title}

    <Original Title Language ("Please Select", ..., "Other")> (A)    ( FIXME: Disambiguate "Title". Someone might think "lyrics" )

    ? (A) != "English":
        [Non-English]
        □ "The artist has provided an English translation"      ( Skip if user.knows_language(lang) )
            [English]
        [Romanized]                                             ( Un-grey if user.has_read_romanization_guide )

    ? (A) == "English":
        [English]

    /------------
    | % MSG Short description and wiki URL
    | * ? (A) != "Please select" -> Un-grey the rest of the group
    | * ? BAD_TITLE_PCRE -> MSG
    | @ ? BAD_TITLE_PCRE -> REP, DRF
    | ! All are blank
    \------------

    Example: BAD_TITLE_PCRE = /-P$|\bf(ea)?t[. ]|【|】|\b(cover|remix)\b/
```
```
{Producers}

    [Producers]

    /------------
    | % MSG http://wiki.vocadb.net/wiki/21/artist-roles
    | ! No producer added || {Intro} says there are no vocaloids
    \------------
```
```
{Vocaloid/UTAU}

    [Vocaloids]    ( Return only humans if {Intro} says there are no vocaloids )

    ▣ "I am not 100% sure about the vocaloid versions used" (A)

    /------------
    | @ ? (A) -> DRF, REP
    | ! No vocalist added
    \------------
```
```
{PVs}

    ◉ "A PV is available" (A)
        [Original PV]
        [Reprint PV]
    ○ "No PV is available" (B)

    /------------
    | * ? (B) -> MSG cases where lack of PV is OK
    | @ ? (B) -> DRF, REP
    | ! (A) checked && URL missing / duplicate.
    \------------
```
```
{Type}

    <Song Type ("Not sure", ...)> (A)

    ? (A) in ("Cover", "Remix"):
        [Original Song]                                         ( Grey if (B) )
        □ "The original song is out of scope" (B)

    /------------
    | * MSG http://wiki.vocadb.net/wiki/6/song-types
    | ? (A) == "Not sure" || (B) -> DRF, REP
    \------------
```
```
{Draft}

    □ Draft
```

### PR DESCRIPTION
**Pull request by Pyther99**
Depends on #817.
Closes #305.
Closes #14.
Partially implements #368. (Romanization link)

Implementation of @dset0x's patch for the song creation page. 

- [x] This PR needs to be merged after #817 to include the new alert
- [x]  The current alert texts need to be split 
- [x] The new alert texts need to be translated

Needs to be tested on the beta site.

### CODE REVIEW

File: VocaDbWeb/wwwroot/Content/Styles/base.less [hunk 1]
```diff
@@ -74,23 +74,23 @@ body {
...
+		margin-left: 1%;
 		width: 84%;
-	}
+	}
 	@media (max-width: @left-menu-breakpoint) {
 		margin-left: 0;
 		margin-right: 0;
 		width: 100%;
-	}
+	}
```
Comment (Pyther99) [inline VocaDbWeb/wwwroot/Content/Styles/base.less:93]: These line endings were still in LF

File: VocaDbWeb/wwwroot/Content/Styles/base.less [hunk 2]
```diff
@@ -166,8 +166,8 @@ body .container-fluid {
 	}
 }
 
-.vdb .page-title {
-	margin-bottom: 10px;
+.vdb .page-title {
+	margin-bottom: 10px;
```
Comment (Pyther99) [inline VocaDbWeb/wwwroot/Content/Styles/base.less:170]: These were also in LF

File: VocaDbWeb/wwwroot/Content/Styles/base.less [hunk 3]
```diff
@@ -345,6 +345,12 @@ body .container-fluid {
 	}
 }
 
+.vcenter {
+	min-height: 100%;
+	min-height: 100vh;
+	display: flex;
+	align-items: center;
+}
```
Comment (Pyther99) [inline VocaDbWeb/wwwroot/Content/Styles/base.less:353]: Could be removed by upgrading bootstrap

File: VocaDbWeb/Views/Song/Create.cshtml [hunk 4]
```diff
@@ -16,156 +16,176 @@
...
-			</div>
-			<div class="editor-field">
-				@Html.ValidationMessage("Artists")
-				<table data-bind="foreach: artists">
-					<tbody>
 						<tr>
+							<td class="formLabel">
+								<a style="text-decoration: underline dashed">
+									@Html.LabelFor(m => m.NameRomaji, htmlAttributes: new { onclick = "window.open('https://wiki.vocadb.net/SiteSettings/Wiki/Index/20?title=Romanization-walkthrough')" })
+								</a>
```
Comment (Pyther99) [inline VocaDbWeb/Views/Song/Create.cshtml:84]: Probably needs a better implementation

**Comment by Pyther99** [comment]
Screenshot:
image

**Comment by Pyther99** [comment]
localhost_44398_Song_Create (1)

**Comment by Pyther99** [comment]
localhost_44398_Song_Create

File: VocaDbWeb/Views/Shared/KnockoutPartials/_CoverArtistMessage.cshtml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿<!--
+	Shows a message for cover artists in an original song.
+	Binding context:  KnockoutComputed<dc.ArtistContract[]>
+-->
+<div class="alert" style="display:none" data-bind="visible: coverArtists() && coverArtists().length > 0 && !canHaveOriginalVersion()">
+	<partial name="Partials/Shared/_NotificationIcon" />
+	@ViewRes.EntryCreateStrings.FoundCoverArtist
```
Comment (Pyther99) [inline VocaDbWeb/Views/Shared/KnockoutPartials/_CoverArtistMessage.cshtml:7]: Causes the AppVeyor build to fail. Will be resolved after the changes from #817 get merged into this PR.

**Comment by ycanardeau** [comment]
@Pyther99 Deployed on the beta site.

**Comment by dset0x** [comment]
@Pyther99 Fantastic work turning the mockup into a functioning page already! It's so cool to see this on the beta site.

I wrote some feedback in this screenshot,
song_submit_feedback

**Comment by Pyther99** [comment]
@dset0x  The markup parsing error occurred because the description was missing a space between ### and the text.

I have fixed the side warnings to the vertical center of the viewport, but this can cause the warnings to overlap if there are enough duplicates. (This worsened the responsiveness for mobile devices and needs to be fixed in the future, see #6)

I also implemented a Bootstrap tooltip for the artist warning.

**Comment by dset0x** [comment]
@Pyther99 That's awesome, thanks for tending to these.

If it worked better on mobile without the fixed location, a workaround could be to only enable that using `@media`, but I can't say how bad that might be.

Yet another minor thing. Not everyone is familiar with the initials "PV". "Promotional Video" might also be misleading though.

**Comment by hikaru-soeda** [comment]
@Pyther99 I think the width of the label in the title input field should be increased.
It doesn't look good in the Japanese UI because the labels are vertical.
Web キャプチャ_9-4-2021_195452_beta vocadb net_

**Comment by dset0x** [comment]
Looking at the current beta site, I it feels a bit awkwardly long now,
image
But if it needs to be so that translations fit better, then it's no issue

**Comment by Pyther99** [comment]
@dset0x The width was reduced in the last commit, which is not yet deployed on the beta site. Nevertheless, I still shortened the width a little.
localhost_44398_Song_Create (3)

**Comment by ycanardeau** [comment]
Thanks for the PR!

**Pull request merged**